### ISSUE TITLE
NVDA does not read content on second Desktop

### ISSUE THREAD

**Issue by yolpsoftware** [nvaccess/nvda#14395]
### Steps to reproduce:

* Start NVDA.
* Execute a software application that uses Windows' `CreateDesktop` API (defined in `user32.dll`). See below for examples.

### Actual behavior:

NVDA does not read the contents on the second Desktop.

### Expected behavior:

NVDA should read the contents of the second Desktop.

#### NVDA installed/portable/running from source:

installed

#### NVDA version:

2022.3.2.27117

#### Windows version:

Windows 10 Home, 10.0.19044 Build 19044

#### Name and version of other software in use when reproducing the issue:

Applications that use the `CreateDesktop` Windows API, see below.

#### Does the issue still occur after restarting your computer?

yes

#### Have you tried any other versions of NVDA? If so, please report their behaviors.

no

#### If NVDA add-ons are disabled, is your problem still occurring?

no add-ons installed

#### Does the issue still occur after you run the COM Registration Fixing Tool in NVDA's tools menu?

yes

#### Safe Exam Browser

The problem has first been identified while running the Safe Exam Browser application (https://safeexambrowser.org/):

* Download and install the latest Safe Exam Browser version
* Run the "SEB Configuration Tool". In the "Security Tab", enable "Create new desktop" under "Kiosk Mode". Save the Settings ("File" --> "Save Settings").
* Close any running Messengers or Video Call applications (or add exceptions under "Applications")
* Start NVDA
* Start Safe Exam Browser
* NVDA will not read the contents
* To quit, click the button in the bottom right corner

Safe Exam Browser will create a new Desktop, where its main window will appear. However, the contents of the window are not read by NVDA. 

#### Separate CreateDesktop Tool

To make sure the problem was not limited to SEB, we tested it with a separate tool.

WARNING: you should run this tool in a virtual machine, or save your work before running it, because it might require you to hard-reboot your PC.

* Download the C++ file here https://github.com/MalwareTech/CreateDesktop/blob/master/Main.cpp
* Compile it to an EXE file. If you use Visual Studio, you might need to change line 41 to `hidden_desktop = CreateHiddenDesktop((char*)"MalwareTech_Hidden");`
* Start NVDA
* Run this tool (WARNING: run this in a VM, or save your work before running the tool - the tool does not offer an option to go back to the previous desktop)

The tool will invoke `CreateDesktop` to create a new desktop. However, the contents of the new desktop are not read by NVDA.

**Comment by michaelDCurran** [comment]
At very least, you would need to some how launch NVDA on the second desktop in order to read its content. We could have further discussion as to whether or not NVDA should detect this scenario and automatically start itself on the second desktop.

**Comment by michaelDCurran** [comment]
I don't follow how sharing the active desktop with Zoom is related to 
running an app on a second desktop. Are you saying that when you do 
this, Zoom creates a second desktop?

It is simply a fact of the Windows architecture that in order for a 
process to interact with windows through window messages and the like, 
it must be running on the same desktop as those windows. So therefore if 
a second desktop is created and made active, NVDA would still need to be 
some how started on that second desktop.
The question then is whether or not it is appropriate for an existing 
NVDA to detect this and force itself to launch on the second desktop. 
Especially taking into account a major reason a second desktop was used 
was probably to isolate a process into a controlled environment.
This is probably what NVDA should do, but we do need to carefully think 
through all the ramifications.

**Comment by yolpsoftware** [comment]
> I don't follow how sharing the active desktop with Zoom is related to running an app on a second desktop. Are you saying that when you do this, Zoom creates a second desktop?

No. What I meant was, if I run a Zoom call, then share my desktop to my "Zoommates", and then create a second desktop, Zoom will automatically share the contents of the second desktop, because that is where the user now "is".

**Comment by michaelDCurran** [comment]
thanks for the clarification.

In that case there is clearly an expectation for NVDA to automatically 
launch on a second desktop if it becomes the active desktop.

However as I previously stated, we will have to think about security 
expectations and how this fits into our priorities.

**Comment by seanbudd** [comment]
Hi @yolpsoftware,

Based on my testing, NVDA works fine on the second desktop.
The issue I noticed is that NVDA does not automatically start on the second desktop.
How to make NVDA automatically start is not immediately clear, nor does it necessarily make sense as a UX for NVDA.

Unfortunately, `ctrl+alt+n` also does not work to start NVDA, which definitely **should** be supported for NVDA.
This is assumed to be a Windows bug, as according to the API, shortcuts in the special folder "AllUsersDesktop" are meant to work on **ALL** desktops.
We plan on reporting this bug to Microsoft.

In the meantime, to start NVDA on the second desktop you can:
- programmatically launch NVDA if it is installed
- manually navigate to the desktop shortcut / start menu item
- manually lock the screen, causing NVDA to automatically start if the relevant setting is enabled

**Comment by yolpsoftware** [comment]
@seanbudd in which way are you creating the second desktop? Windows offers two ways:

a) the Virtual Desktop you create by pressing Win + Ctrl + d, or by going to Task View --> New Desktop.
b) the second desktop you create by using the `CreateDesktop` Win32 API.

These are not the same, and this ticket is about case b). In all our tests, NVDA did not read the contents from the second desktop.

**Comment by seanbudd** [comment]
@yolpsoftware 

I am talking about case b. Can you confirm you tried the following work around steps?

- programmatically launch NVDA if it is installed
- manually navigate to the desktop shortcut / start menu item, launch NVDA
- manually lock the screen, causing NVDA to automatically start if the relevant setting is enabled

**Comment by yolpsoftware** [comment]
@seanbudd not sure I understand. This is what I do:

* Launch NVDA normally (e.g. through start menu)
* Launch a software that uses the `CreateDesktop` Win32 API (see C++ tool in my original post, please read warning there)
* Second desktop opens
* NVDA does not read the contents of the second desktop

Please compare this to the behavior of other software (like Zoom or screen recorders), which automatically record the one desktop where the user currently "is".

Do you mean we need to programmatically launch NVDA on the second desktop?

**Comment by seanbudd** [comment]
@yolp please read my earlier reply.

Can you confirm that NVDA works when you navigate to the desktop shortcut or start menu item and launch it?

I listed 3 workarounds, programmatically launching is one of them.

**Comment by Adriani90** [comment]
@yolpsoftware are you saying that for example with Narator or Jaws started on the first desktop you can read and navigate the content on the second desktop without a screen reader session started on the second desktop?

I am trying to understand if this is NVDA specific issue or if it happens with Jaws and Narator as well.

**Comment by seanbudd** [comment]
we've updated the ticket to focus on the external bug with windows that `ctrl+alt+n` doesn't launch NVDA on the second desktop

**Comment by Adriani90** [comment]
@yolpsoftware when you create a new desktop that way, does the application create also a new mouse cursor / system focus / system caret? Or are these the same cursors as in desktop 1?
I am asking because if desktop 2 is in foreground and it has programatically another cursor than desktop 1, then NVDA cannot get the focused elements since it is tied to the system focus of desktop 1. So it cannot detect whether you change the cursor position on desktop 2.

Does the new NVDA remote feature help in this case?
cc: @SaschaCowley
